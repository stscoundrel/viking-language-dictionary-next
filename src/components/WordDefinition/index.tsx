import { capitalize } from 'lib/utils/strings'
import { lettersToShortTwigRunes, lettersToLongBranchRunes } from 'younger-futhark'
import { Abbreviation, addAbbreviationsToContent } from 'lib/services/abbreviations'
import { DictionaryEntry } from 'lib/models/dictionary'
import Abbreviations from 'components/Abbreviations'
import styles from './WordDefinition.module.scss'

interface WordDefinitionProps{
  data: DictionaryEntry,
  abbreviations: Abbreviation[]
}

export default function WordDefinition({ data, abbreviations }: WordDefinitionProps) {
  const { headword, definition } = data

  return (
    <article className={styles.section}>
      <header>
        <h1 lang="non">{capitalize(headword)}</h1>

        <small className={styles.subHeading}>
          Viking Language Dictionary - {headword.toLowerCase()}
        </small>
        <p>Meaning of Old Norse / Old Icelandic word <em>&quot;{headword}&quot;</em></p>
      </header>

      <dl className={styles.definitionList}>
        <dt><strong>{headword}</strong></dt>
        <dd
          lang="nor"
          className={styles.itemDescription}
          dangerouslySetInnerHTML={{
            __html: addAbbreviationsToContent(definition, abbreviations),
          } }
        ></dd>
      </dl>

      <p>Possible runic inscriptions in <em>Younger Futhark:</em></p>
      <ul>
        <li>
          - Long branch (Danish):
          <span className={styles.rune}>{ lettersToLongBranchRunes(headword) }</span>
        </li>
        <li>
          - Short twig (Swedish & Norwegian):
          <span className={styles.rune}>{ lettersToShortTwigRunes(headword) }</span>
        </li>
      </ul>

      <Abbreviations abbreviations={abbreviations} />
    </article>
  )
}
