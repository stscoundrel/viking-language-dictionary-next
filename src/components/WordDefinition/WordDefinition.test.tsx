import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import WordDefinition from './index'

const word = {
  headword: 'Þiðrandi',
  definition: 'm Thidrandi (personal name)',
  isCommon: false,
  slug: 'thidrandi',
}

const abbreviations = [
  {
    abbreviation: 'n.',
    explanation: 'Neutrum.',
  },
  {
    abbreviation: 's.',
    explanation: 'substantiv.',
  },
]

describe('WordDefinition component', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<WordDefinition data={word} abbreviations={abbreviations} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <WordDefinition data={word} abbreviations={abbreviations} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct label', () => {
    const tree = renderer.create(<WordDefinition data={word} abbreviations={abbreviations} />)
    const { root } = tree

    expect(root.findByType('h1').children).toEqual(['Þiðrandi'])
  })
})
