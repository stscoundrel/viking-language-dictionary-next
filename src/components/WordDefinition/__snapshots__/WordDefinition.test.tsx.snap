// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordDefinition component Matches snapshot 1`] = `
<article
  className="section"
>
  <header>
    <h1
      lang="non"
    >
      Þiðrandi
    </h1>
    <small
      className="subHeading"
    >
      Viking Language Dictionary - 
      þiðrandi
    </small>
    <p>
      Meaning of Old Norse / Old Icelandic word 
      <em>
        "
        Þiðrandi
        "
      </em>
    </p>
  </header>
  <dl
    className="definitionList"
  >
    <dt>
      <strong>
        Þiðrandi
      </strong>
    </dt>
    <dd
      className="itemDescription"
      dangerouslySetInnerHTML={
        Object {
          "__html": "m Thidrandi (personal name)",
        }
      }
      lang="nor"
    />
  </dl>
  <p>
    Possible runic inscriptions in 
    <em>
      Younger Futhark:
    </em>
  </p>
  <ul>
    <li>
      - Long branch (Danish):
      <span
        className="rune"
      >
        ᚦᛁᚦᚱᛅᚾᛏᛁ
      </span>
    </li>
    <li>
      - Short twig (Swedish & Norwegian):
      <span
        className="rune"
      >
        ᚦᛁᚦᚱᛆᚿᛐᛁ
      </span>
    </li>
  </ul>
  <div
    className="abbreviations"
  >
    <div
      className="column"
    >
      <h4>
        Abbreviations used:
      </h4>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            n.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          Neutrum.
        </dd>
      </dl>
      <dl
        className="wrap"
      >
        <dt
          className="abbreviation"
        >
          <strong>
            s.
          </strong>
        </dt>
        <dd
          className="explanation"
        >
          substantiv.
        </dd>
      </dl>
    </div>
  </div>
</article>
`;
